### PR TITLE
warn user if he leaves profile editing page with unsaved changes #6520

### PR DESCRIPTION
**Pull request by atulonruby**

### CODE REVIEW

File: app/assets/javascripts/app/app.js [hunk 1]
```diff
@@ -170,5 +170,6 @@ var app = {
 
 $(function() {
   app.initialize();
+  $('form').areYouSure();
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/app.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+ * Version: 1.9.0
+ * Date:    13th August 2014
+ */
+(function($) {
+  
+  $.fn.areYouSure = function(options) {
+      
+    var settings = $.extend(
+      {
+        'message' : 'You have unsaved changes!',
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        'dirtyClass' : 'dirty',
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        'change' : null,
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        'silent' : false,
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        'addRemoveFieldsMarksDirty' : false,
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        'fieldEvents' : 'change keyup propertychange input',
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        'fieldSelector': ":input:not(input[type=submit]):not(input[type=button])"
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+      }, options);
+
+    var getValue = function($field) {
+      if ($field.hasClass('ays-ignore')
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+          || $field.hasClass('aysIgnore')
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Bad line breaking before '||'.

Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+          || $field.attr('data-ays-ignore')
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Bad line breaking before '||'.

Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+          || $field.attr('name') === undefined) {
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Bad line breaking before '||'.

Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        return null;
+      }
+
+      if ($field.is(':disabled')) {
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+        return 'ays-disabled';
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+      }
+
+      var val;
+      var type = $field.attr('type');
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+      if ($field.is('select')) {
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+        type = 'select';
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+      }
+
+      switch (type) {
+        case 'checkbox':
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+        case 'radio':
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+          val = $field.is(':checked');
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+          break;
+        case 'select':
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+          val = '';
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+          $field.find('option').each(function(o) {
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+            var $option = $(this);
+            if ($option.is(':selected')) {
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+          break;
+        default:
+          val = $field.val();
+      }
+
+      return val;
+    };
+
+    var storeOrigValue = function($field) {
+      $field.data('ays-orig', getValue($field));
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+    };
+
+    var checkForm = function(evt) {
+
+      var isFieldDirty = function($field) {
+        var origValue = $field.data('ays-orig');
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+        if (undefined === origValue) {
+          return false;
+        }
+        return (getValue($field) != origValue);
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Expected '!==' and instead saw '!='.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+      };
+
+      var $form = ($(this).is('form'))
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+      var isFieldDirty = function($field) {
+        var origValue = $field.data('ays-orig');
+        if (undefined === origValue) {
+          return false;
+        }
+        return (getValue($field) != origValue);
+      };
+
+      var $form = ($(this).is('form')) 
+                    ? $(this)
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Bad line breaking before '?'.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+                    : $(this).parents('form');
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Strings must use doublequote.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 31]
```diff
@@ -0,0 +1,192 @@
...
+        setDirtyStatus($form, true);
+        return;
+      }
+
+      $fields = $form.find(settings.fieldSelector);
+
+      if (settings.addRemoveFieldsMarksDirty) {              
+        // Check if field count has changed
+        var origCount = $form.data("ays-orig-field-count");
+        if (origCount != $fields.length) {
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Expected '!==' and instead saw '!='.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 32]
```diff
@@ -0,0 +1,192 @@
...
+      var fields = $form.find(settings.fieldSelector);
+      $(fields).each(function() { storeOrigValue($(this)); });
+      $(fields).unbind(settings.fieldEvents, checkForm);
+      $(fields).bind(settings.fieldEvents, checkForm);
+      $form.data("ays-orig-field-count", $(fields).length);
+      setDirtyStatus($form, false);
+    };
+
+    var setDirtyStatus = function($form, isDirty) {
+      var changed = isDirty != $form.hasClass(settings.dirtyClass);
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Expected '!==' and instead saw '!='.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 33]
```diff
@@ -0,0 +1,192 @@
...
+      $form.toggleClass(settings.dirtyClass, isDirty);
+        
+      // Fire change event if required
+      if (changed) {
+        if (settings.change) settings.change.call($form, $form);
```
Comment (diaspora-code-review) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Expected '{' and instead saw 'settings'.

File: lib/assets/javascripts/jquery.are-you-sure.js [hunk 34]
```diff
@@ -0,0 +1,192 @@
+/*!
+ * jQuery Plugin: Are-You-Sure (Dirty Form Detection)
+ * https://github.com/codedance/jquery.AreYouSure/
```
Comment (jhass) [inline lib/assets/javascripts/jquery.are-you-sure.js]: Please add `gem "rails-assets-jquery.are-you-sure", "1.9.0"` to the `Gemfile` instead.

File: app/assets/javascripts/app/app.js (hunk 1 above)
Comment (jhass) [inline app/assets/javascripts/app/app.js]: I'm not sure we should do this for every single form, especially search boxes come to mind where it's unwanted.

Also if we're adding a plugin for this, we probably should replace the existing custom code for the publisher with it.

File: lib/assets/javascripts/jquery.are-you-sure.js (hunk 2 above)
Comment (jhass) [inline lib/assets/javascripts/jquery.are-you-sure.js]: If you want to use the plugin, you'll have to hook these into our i18n system.

File: app/assets/javascripts/app/app.js [hunk 35]
```diff
@@ -170,5 +170,6 @@ var app = {
 
 $(function() {
   app.initialize();
+  $("#update_profile_form").areYouSure();
```
Comment (jhass) [inline app/assets/javascripts/app/app.js]: I'm not sure this the best location for this specific call. I think https://github.com/diaspora/diaspora/blob/develop/app/assets/javascripts/app/pages/settings.js is hooked up to the profile edit page?

**Comment by atulonruby** [comment]
@jhass : This merge commit is not reflecting the change on the develop branch.
Can you please check it

**Comment by jhass** [comment]
@atulonruby mmh, did you maybe push a clean upstream develop branch onto the branch this PR comes from?

This is one reason we ask you to do feature branches.

**Comment by atulonruby** [comment]
I have sent another pull request from a different branch (feature branch). please accept
https://github.com/diaspora/diaspora/pull/6589

**Pull request merged**